gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java

```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
     writeAndVerifyTemporaryFiles(resourceId, 1);
   }
 
+  @Test
+  public void uploadViaPCUVerifyPartFileCleanup() throws IOException, InterruptedException {
+    String partFilePrefix = name.getMethodName();
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                pcuDefaultOptions.toBuilder().setPartFilePrefix(partFilePrefix).build())
+            .build();
+
+    gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+
+    writeAndVerifyPartFiles(bufferCapacity, resourceId, partFileCount, 0, partFilePrefix);
+  }
+
+  @Test
+  public void uploadViaPCUVerifyPartFileNotCleanedUp() throws IOException, InterruptedException {
+    String partFilePrefix = name.getMethodName();
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                pcuDefaultOptions.toBuilder()
+                    .setPartFilePrefix(partFilePrefix)
+                    .setPartFileCleanupType(PartFileCleanupType.NEVER)
+                    .build())
+            .build();
+
+    gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+    // part file not cleaned up because PartFileCleanupType.NEVER is used.
+    writeAndVerifyPartFiles(
+        bufferCapacity, resourceId, partFileCount, partFileCount, partFilePrefix);
+  }
+
+  @Test
+  public void uploadViaPCUComposeFileMissingFailure() throws IOException, InterruptedException {
+    String partFilePrefix = name.getMethodName();
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                pcuDefaultOptions.toBuilder().setPartFilePrefix(partFilePrefix).build())
+            .build();
+
+    gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+
+    byte[] bytesToWrite = new byte[partFileCount * bufferCapacity];
+    GoogleCloudStorageTestHelper.fillBytes(bytesToWrite);
+    WritableByteChannel writeChannel = gcs.create(resourceId);
+    writeChannel.write(ByteBuffer.wrap(bytesToWrite));
+    // part files are getting uploaded in async thread
+    // wait for it to complete before listing files
+    Thread.sleep(5000);
```

Comment:
This should be avoided. This will increase our test run time. Wont' close() delete this?